singer_sdk/sinks/sql.py

```diff
@@ -65,36 +61,76 @@ def connection(self) -> sqlalchemy.engine.Connection:
 
     @property
     def table_name(self) -> str:
-        """Returns the table name, with no schema or database part.
+        """Return the table name, with no schema or database part.
 
         Returns:
             The target table name.
         """
         parts = self.stream_name.split("-")
-
-        if len(parts) == 1:
-            return self.stream_name
-        else:
-            return parts[-1]
+        return self.stream_name if len(parts) == 1 else parts[-1]
 
     @property
     def schema_name(self) -> Optional[str]:
-        """Returns the schema name or `None` if using names with no schema part.
+        """Return the schema name or `None` if using names with no schema part.
 
         Returns:
             The target schema name.
         """
-        return None  # Assumes single-schema target context.
+        parts = self.stream_name.split("-")
+        if len(parts) == 2:
+            return parts[0]
+        if len(parts) == 3:
+            return parts[1]
+        return None
 
     @property
     def database_name(self) -> Optional[str]:
-        """Returns the DB name or `None` if using names with no database part.
+        """Return the DB name or `None` if using names with no database part.
 
         Returns:
             The target database name.
         """
         return None  # Assumes single-DB target context.
 
+    @property
+    def full_table_name(self) -> str:
+        """Return the fully qualified table name.
+
+        Returns:
+            The fully qualified table name.
+        """
+        return self.connector.get_fully_qualified_name(
+            table_name=self.table_name,
+            schema_name=self.schema_name,
+            db_name=self.database_name,
+        )
+
+    @property
+    def full_schema_name(self) -> str:
+        """Return the fully qualified schema name.
+
+        Returns:
+            The fully qualified schema name.
+        """
+        return self.connector.get_fully_qualified_name(
+            schema_name=self.schema_name, db_name=self.database_name
+        )
+
+    def setup(self) -> None:
+        """Set up Sink.
+
+        This method is called on Sink creation, and creates the required Schema and
+        Table entities in the target database.
+        """
+        if self.schema_name:
+            self.connector.prepare_schema(self.schema_name)
```

Comment:
Should `full_schema_name` be used here?

Reply:
@edgarrmondragon This isn't super clear from the documentation but SQLAlchemy doesn't appear to have native support for fully-qualified names 🤔 e.g. when specifying the schema for a table, you are supposed to do it either on the `MetaData()` metadata object (as I am doing [here](https://github.com/meltano/sdk/pull/1036/files#diff-61b20de8a12c7c91dfc4de62be5a230d2b4b1b2dd688628705554dc63a9490b6R596)) which will apply to all tables used with that metadata, or on the `Table()` object itself (as per the docs example [here](https://docs.sqlalchemy.org/en/14/core/metadata.html#specifying-the-schema-name)):

```python
metadata_obj = MetaData()

financial_info = Table(
    "financial_info",
    metadata_obj,
    Column("id", Integer, primary_key=True),
    Column("value", String(100), nullable=False),
    schema="remote_banks",
)
```

From my testing, I found that passing a fully qualified name (for table or schema) would result in a quoted identifier (as SQLAlchemy correctly detects and escapes the illegal `.` char in the identifier). I.e. in stead of creating an example fully-qualified table of `database.schema.table`, SQLAlchemy would create a table called `"database.schema.table"` in the default/public schema of the current database (from the connection object) 🤦‍♂️

This is different to executing raw SQL directly, where fully-qualified names (when supported by the underlying database) are an option (inserted without quotes). We may prefer to just compose the SQL ourselves and make the composition overridable, but I implemented via `CreateSchema` on the basis that SQLAlchemy will then do the work to compose dialect-specific SQL for us 👍

Reply:
> From my testing, I found that passing a fully qualified name (for table or schema) would result in a quoted identifier (as SQLAlchemy correctly detects and escapes the illegal . char in the identifier).

That makes sense 👍. Let's stick with the `Metadata` defining the namespace and SQLAlchemy generating the right schema identifier.